File: implementation/scipamato/scipamato-core-sync/src/test/java/ch/difty/scipamato/core/sync/jobs/paper/SyncShortFieldWithEmptyMainFieldConcatenatorTest.java
package ch.difty.scipamato.core.sync.jobs.paper;

import static ch.difty.scipamato.core.db.public_.tables.Paper.PAPER;
import static org.assertj.core.api.AssertionsForInterfaceTypes.assertThat;
import static org.mockito.Mockito.*;

import java.sql.ResultSet;
import java.sql.SQLException;

import org.jooq.TableField;
import org.junit.jupiter.api.AfterEach;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.Mock;
import org.mockito.junit.jupiter.MockitoExtension;

import ch.difty.scipamato.common.TestUtils;
import ch.difty.scipamato.core.db.public_.tables.records.PaperRecord;

@ExtendWith(MockitoExtension.class)
public class SyncShortFieldWithEmptyMainFieldConcatenatorTest {

    private final SyncShortFieldConcatenator sfc = new SyncShortFieldWithEmptyMainFieldConcatenator();

    @Mock
    private ResultSet resultSet;

    @AfterEach
    public void tearDown() {
        verifyNoMoreInteractions(resultSet);
    }

    @Test
    public void methods_withNullRecordset_throws() {
        TestUtils.assertDegenerateSupplierParameter(() -> sfc.methodsFrom(null), "rs");
    }

    @Test
    public void population_withNullRecordset_throws() {
        TestUtils.assertDegenerateSupplierParameter(() -> sfc.populationFrom(null), "rs");
    }

    @Test
    public void result_withNullRecordset_throws() {
        TestUtils.assertDegenerateSupplierParameter(() -> sfc.resultFrom(null), "rs");
    }

    @Test
    void methods_withNonNullMethod_returnsMethod() throws SQLException {
        stubMethodFieldsWithMainFieldReturning("method");
        assertThat(sfc.methodsFrom(resultSet)).isEqualTo("method");
        verifyCallingMethodsFields();
    }

    @Test
    void methods_withNullMethod_returnsConcatenatedShortMethodFieldsConcatenated() throws SQLException {
        stubMethodFieldsWithMainFieldReturning(null);
        assertThat(sfc.methodsFrom(resultSet)).isEqualTo(
            "Study Design: msd / Outcome: mo / Place: pp / Pollutant: ep / Exposure Assessment: ea / Statistical Method: ms / Confounders: mc");
        verifyCallingMethodsFields();
    }

    private void stubMethodFieldsWithMainFieldReturning(final String mainFixture) throws SQLException {
        when(resultSet.getString(anyString())).thenAnswer(invocationsOnMock -> {
            String fieldName = (String) invocationsOnMock.getArguments()[0];
            if (fieldName.equals(PAPER.METHODS.getName()))
                return mainFixture;
            else if (fieldName.equals(PAPER.METHOD_STUDY_DESIGN.getName()))
                return "msd";
            else if (fieldName.equals(PAPER.METHOD_OUTCOME.getName()))
                return "mo";
            else if (fieldName.equals(PAPER.POPULATION_PLACE.getName()))
                return "pp";
            else if (fieldName.equals(PAPER.EXPOSURE_POLLUTANT.getName()))
                return "ep";
            else if (fieldName.equals(PAPER.EXPOSURE_ASSESSMENT.getName()))
                return "ea";
            else if (fieldName.equals(PAPER.METHOD_STATISTICS.getName()))
                return "ms";
            else if (fieldName.equals(PAPER.METHOD_CONFOUNDERS.getName()))
                return "mc";
            else
                throw new IllegalArgumentException("unhandled stub");
        });
    }

    private void verifyCallingMethodsFields() throws SQLException {
        verify(resultSet).getString(PAPER.METHODS.getName());
        verify(resultSet).getString(PAPER.METHOD_STUDY_DESIGN.getName());
        verify(resultSet).getString(PAPER.METHOD_OUTCOME.getName());
        verify(resultSet).getString(PAPER.POPULATION_PLACE.getName());
        verify(resultSet).getString(PAPER.EXPOSURE_POLLUTANT.getName());
        verify(resultSet).getString(PAPER.EXPOSURE_ASSESSMENT.getName());
        verify(resultSet).getString(PAPER.METHOD_STATISTICS.getName());
        verify(resultSet).getString(PAPER.METHOD_CONFOUNDERS.getName());
    }

    @Test
    void population_withNonNullPopulation_returnsPopulation() throws SQLException {
        stubPopulationFieldsWithMainFieldReturning("population");
        assertThat(sfc.populationFrom(resultSet)).isEqualTo("population");
        verifyCallingPopulationFields();
    }

    @Test
    void population_withNullPopulation_returnsPopulationShortFieldsConcatenated() throws SQLException {
        stubPopulationFieldsWithMainFieldReturning(null);
        assertThat(sfc.populationFrom(resultSet)).isEqualTo("Place: ppl / Participants: ppa / Study Duration: pd");
        verifyCallingPopulationFields();
    }

    private void stubPopulationFieldsWithMainFieldReturning(final String mainFixture) throws SQLException {
        when(resultSet.getString(anyString())).thenAnswer(invocationsOnMock -> {
            String fieldName = (String) invocationsOnMock.getArguments()[0];
            if (fieldName.equals(PAPER.POPULATION.getName()))
                return mainFixture;
            else if (fieldName.equals(PAPER.POPULATION_PLACE.getName()))
                return "ppl";
            else if (fieldName.equals(PAPER.POPULATION_PARTICIPANTS.getName()))
                return "ppa";
            else if (fieldName.equals(PAPER.POPULATION_DURATION.getName()))
                return "pd";
            else
                throw new IllegalArgumentException("unhandled stub");
        });
    }

    private void verifyCallingPopulationFields() throws SQLException {
        verify(resultSet).getString(PAPER.POPULATION.getName());
        verify(resultSet).getString(PAPER.POPULATION_PLACE.getName());
        verify(resultSet).getString(PAPER.POPULATION_PARTICIPANTS.getName());
        verify(resultSet).getString(PAPER.POPULATION_DURATION.getName());
    }

    @Test
    void result_withNonNullResult_returnsResult() throws SQLException {
        stubResultFieldsWithMainFieldReturning("result");
        assertThat(sfc.resultFrom(resultSet)).isEqualTo("result");
        verifyCallingResultFields();
    }

    @Test
    void result_withNullResult_returnsResultShortFieldsConcatenated() throws SQLException {
        stubResultFieldsWithMainFieldReturning(null);
        assertThat(sfc.resultFrom(resultSet)).isEqualTo(
            "Measured Outcome: rmo / Exposure (Range): rer / Effect Estimate: ree / Conclusion: cc");
        verifyCallingResultFields();
    }

    private void stubResultFieldsWithMainFieldReturning(final String mainFixture) throws SQLException {
        when(resultSet.getString(anyString())).thenAnswer(invocationsOnMock -> {
            String fieldName = (String) invocationsOnMock.getArguments()[0];
            if (fieldName.equals(PAPER.RESULT.getName()))
                return mainFixture;
            else if (fieldName.equals(PAPER.RESULT_MEASURED_OUTCOME.getName()))
                return "rmo";
            else if (fieldName.equals(PAPER.RESULT_EXPOSURE_RANGE.getName()))
                return "rer";
            else if (fieldName.equals(PAPER.RESULT_EFFECT_ESTIMATE.getName()))
                return "ree";
            else if (fieldName.equals(PAPER.CONCLUSION.getName()))
                return "cc";
            else
                throw new IllegalArgumentException("unhandled stub");
        });
    }

    private void verifyCallingResultFields() throws SQLException {
        verify(resultSet).getString(PAPER.RESULT.getName());
        verify(resultSet).getString(PAPER.RESULT_MEASURED_OUTCOME.getName());
        verify(resultSet).getString(PAPER.RESULT_EXPOSURE_RANGE.getName());
        verify(resultSet).getString(PAPER.RESULT_EFFECT_ESTIMATE.getName());
        verify(resultSet).getString(PAPER.CONCLUSION.getName());
    }

    private final SyncShortFieldConcatenator throwingConcatenator = new SyncShortFieldWithEmptyMainFieldConcatenator() {
        @Override
        String methodsFrom(final ResultSet rs, final TableField<PaperRecord, String> methodField,
            final TableField<PaperRecord, String> methodStudyDesignField,
            final TableField<PaperRecord, String> methodOutcomeField,
            final TableField<PaperRecord, String> populationPlaceField,
            final TableField<PaperRecord, String> exposurePollutantField,
            final TableField<PaperRecord, String> exposureAssessmentField,
            final TableField<PaperRecord, String> methodStatisticsField,
            final TableField<PaperRecord, String> methodConfoundersField) throws SQLException {
            throw new SQLException("boom");
        }

        @Override
        String populationFrom(final ResultSet rs, final TableField<PaperRecord, String> populationField,
            final TableField<PaperRecord, String> populationPlaceField,
            final TableField<PaperRecord, String> populationParticipantsField,
            final TableField<PaperRecord, String> populationDurationField) throws SQLException {
            throw new SQLException("boom");
        }

        @Override
        String resultFrom(final ResultSet rs, final TableField<PaperRecord, String> resultField,
            final TableField<PaperRecord, String> resultMeasuredOutcomeField,
            final TableField<PaperRecord, String> resultExposureRangeField,
            final TableField<PaperRecord, String> resultEffectEstimateField,
            final TableField<PaperRecord, String> conclusionField) throws SQLException {
            throw new SQLException("boom");
        }
    };

    @Test
    public void methodsFrom_withThrowingMethod_returnsNull() {
        assertThat(throwingConcatenator.methodsFrom(mock(ResultSet.class))).isNull();
    }

    @Test
    public void polulationFrom_withThrowingMethod_returnsNull() {
        assertThat(throwingConcatenator.populationFrom(mock(ResultSet.class))).isNull();
    }

    @Test
    public void resultFrom_withThrowingMethod_returnsNull() {
        assertThat(throwingConcatenator.resultFrom(mock(ResultSet.class))).isNull();
    }

}

File: core/core-sync/src/test/kotlin/ch/difty/scipamato/core/sync/jobs/paper/SyncShortFieldWithEmptyMainFieldConcatenatorTest.kt
@file:Suppress("SpellCheckingInspection")

package ch.difty.scipamato.core.sync.jobs.paper

import ch.difty.scipamato.core.db.tables.Paper.PAPER
import ch.difty.scipamato.core.db.tables.records.PaperRecord
import io.mockk.confirmVerified
import io.mockk.every
import io.mockk.mockk
import io.mockk.slot
import io.mockk.verify
import org.amshove.kluent.shouldBeEqualTo
import org.amshove.kluent.shouldBeNull
import org.jooq.TableField
import org.junit.jupiter.api.AfterEach
import org.junit.jupiter.api.Test
import java.sql.ResultSet
import java.sql.SQLException

internal class SyncShortFieldWithEmptyMainFieldConcatenatorTest {

    private val sfc = SyncShortFieldWithEmptyMainFieldConcatenator()

    private val resultSet = mockk<ResultSet>()

    private val throwingConcatenator = object : SyncShortFieldWithEmptyMainFieldConcatenator() {
        override fun methodsFrom(
            rs: ResultSet,
            methodField: TableField<PaperRecord, String?>,
            methodStudyDesignField: TableField<PaperRecord, String?>,
            methodOutcomeField: TableField<PaperRecord, String?>,
            populationPlaceField: TableField<PaperRecord, String?>,
            exposurePollutantField: TableField<PaperRecord, String?>,
            exposureAssessmentField: TableField<PaperRecord, String?>,
            methodStatisticsField: TableField<PaperRecord, String?>,
            methodConfoundersField: TableField<PaperRecord, String?>
        ): String = throw SQLException("boom")

        override fun populationFrom(
            rs: ResultSet,
            populationField: TableField<PaperRecord, String?>,
            populationPlaceField: TableField<PaperRecord, String?>,
            populationParticipantsField: TableField<PaperRecord, String?>,
            populationDurationField: TableField<PaperRecord, String?>
        ): String = throw SQLException("boom")

        override fun resultFrom(
            rs: ResultSet,
            resultField: TableField<PaperRecord, String?>,
            resultMeasuredOutcomeField: TableField<PaperRecord, String?>,
            resultExposureRangeField: TableField<PaperRecord, String?>,
            resultEffectEstimateField: TableField<PaperRecord, String?>,
            conclusionField: TableField<PaperRecord, String?>
        ): String = throw SQLException("boom")
    }

    @AfterEach
    fun tearDown() {
        confirmVerified(resultSet)
    }

    @Test
    fun methods_withNonNullMethod_returnsMethod() {
        stubMethodFieldsWithMainFieldReturning("method")
        sfc.methodsFrom(resultSet) shouldBeEqualTo "method"
        verifyCallingMethodsFields()
    }

    @Test
    fun methods_withNullMethod_returnsConcatenatedShortMethodFieldsConcatenated() {
        stubMethodFieldsWithMainFieldReturning(null)
        sfc.methodsFrom(resultSet) shouldBeEqualTo
            "Studiendesign: msd / Zielgrössen: mo / Ort/Land: pp / Schadstoff: ep / " +
            "Belastungsabschätzung: ea / Statistische Methode: ms / Störfaktoren: mc"
        verifyCallingMethodsFields()
    }

    @Suppress("ComplexMethod")
    private fun stubMethodFieldsWithMainFieldReturning(mainFixture: String?) {
        val slot = slot<String>()
        every { resultSet.getString(capture(slot)) } answers {
            when (slot.captured) {
                PAPER.METHODS.name -> mainFixture
                PAPER.METHOD_STUDY_DESIGN.name -> "msd"
                PAPER.METHOD_OUTCOME.name -> "mo"
                PAPER.POPULATION_PLACE.name -> "pp"
                PAPER.EXPOSURE_POLLUTANT.name -> "ep"
                PAPER.EXPOSURE_ASSESSMENT.name -> "ea"
                PAPER.METHOD_STATISTICS.name -> "ms"
                PAPER.METHOD_CONFOUNDERS.name -> "mc"
                else -> throw IllegalArgumentException("unhandled stub")
            }
        }
    }

    private fun verifyCallingMethodsFields() {
        verify { resultSet.getString(PAPER.METHODS.name) }
        verify { resultSet.getString(PAPER.METHOD_STUDY_DESIGN.name) }
        verify { resultSet.getString(PAPER.METHOD_OUTCOME.name) }
        verify { resultSet.getString(PAPER.POPULATION_PLACE.name) }
        verify { resultSet.getString(PAPER.EXPOSURE_POLLUTANT.name) }
        verify { resultSet.getString(PAPER.EXPOSURE_ASSESSMENT.name) }
        verify { resultSet.getString(PAPER.METHOD_STATISTICS.name) }
        verify { resultSet.getString(PAPER.METHOD_CONFOUNDERS.name) }
    }

    @Test
    fun population_withNonNullPopulation_returnsPopulation() {
        stubPopulationFieldsWithMainFieldReturning("population")
        sfc.populationFrom(resultSet) shouldBeEqualTo "population"
        verifyCallingPopulationFields()
    }

    @Test
    fun population_withNullPopulation_returnsPopulationShortFieldsConcatenated() {
        stubPopulationFieldsWithMainFieldReturning(null)
        sfc.populationFrom(resultSet) shouldBeEqualTo "Ort/Land: ppl / Studienteilnehmer: ppa / Studiendauer: pd"
        verifyCallingPopulationFields()
    }

    private fun stubPopulationFieldsWithMainFieldReturning(mainFixture: String?) {
        val slot = slot<String>()
        every { resultSet.getString(capture(slot)) } answers {
            when (slot.captured) {
                PAPER.POPULATION.name -> mainFixture
                PAPER.POPULATION_PLACE.name -> "ppl"
                PAPER.POPULATION_PARTICIPANTS.name -> "ppa"
                PAPER.POPULATION_DURATION.name -> "pd"
                else -> throw IllegalArgumentException("unhandled stub")
            }
        }
    }

    private fun verifyCallingPopulationFields() {
        verify { resultSet.getString(PAPER.POPULATION.name) }
        verify { resultSet.getString(PAPER.POPULATION_PLACE.name) }
        verify { resultSet.getString(PAPER.POPULATION_PARTICIPANTS.name) }
        verify { resultSet.getString(PAPER.POPULATION_DURATION.name) }
    }

    @Test
    fun result_withNonNullResult_returnsResult() {
        stubResultFieldsWithMainFieldReturning("result")
        sfc.resultFrom(resultSet) shouldBeEqualTo "result"
        verifyCallingResultFields()
    }

    @Test
    fun result_withNullResult_returnsResultShortFieldsConcatenated() {
        stubResultFieldsWithMainFieldReturning(null)
        sfc.resultFrom(resultSet) shouldBeEqualTo
            "Gemessene Zielgrösse: rmo / Gemessene Belastung (Spanne): rer / Resultate: ree / Schlussfolgerung: cc"
        verifyCallingResultFields()
    }

    private fun stubResultFieldsWithMainFieldReturning(mainFixture: String?) {
        val slot = slot<String>()
        every { resultSet.getString(capture(slot)) } answers {
            when (slot.captured) {
                PAPER.RESULT.name -> mainFixture
                PAPER.RESULT_MEASURED_OUTCOME.name -> "rmo"
                PAPER.RESULT_EXPOSURE_RANGE.name -> "rer"
                PAPER.RESULT_EFFECT_ESTIMATE.name -> "ree"
                PAPER.CONCLUSION.name -> "cc"
                else -> throw IllegalArgumentException("unhandled stub")
            }
        }
    }

    private fun verifyCallingResultFields() {
        verify { resultSet.getString(PAPER.RESULT.name) }
        verify { resultSet.getString(PAPER.RESULT_MEASURED_OUTCOME.name) }
        verify { resultSet.getString(PAPER.RESULT_EXPOSURE_RANGE.name) }
        verify { resultSet.getString(PAPER.RESULT_EFFECT_ESTIMATE.name) }
        verify { resultSet.getString(PAPER.CONCLUSION.name) }
    }

    @Test
    fun methodsFrom_withThrowingMethod_returnsNull() {
        throwingConcatenator.methodsFrom(mockk()).shouldBeNull()
    }

    @Test
    fun populationFrom_withThrowingMethod_returnsNull() {
        throwingConcatenator.populationFrom(mockk()).shouldBeNull()
    }

    @Test
    fun resultFrom_withThrowingMethod_returnsNull() {
        throwingConcatenator.resultFrom(mockk()).shouldBeNull()
    }
}
